proposals/p1885.md

```diff
@@ -0,0 +1,866 @@
+-   Facilitate code reuse, compared to reimplementing an interface
+-   Direct access to `self`, limiting needs for pointers and address resolution
+
+#### Mutable values
+
+Named mixins, used as programmable members, could expose a view with mutable
+values:
+
+```carbon
+// Using a named mixin to expose mutable values
+for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
+```
+
+Used in this way, this clarifies that this a view of the data, like an
+attribute, with no direct side effects, when compared to a method.
+
+#### Alternative views
+
+Similarly, mixins could allow supporting other views into the container.
+
+For example, maps could expose a view with "key" and "value" only, in addition
+to the default (key, value) tuple.
+
+```carbon
+// Using a named mixin to expose mutable values
+for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
+```
+
+And "reversed" view could also be used for reverse iteration:
+
+```carbon
+// Reversing
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
+
+Large elements would have a negative impact if the `Optional` cannot leverage
+unused bit patterns within the type, nor present a view instead of a copy.
+
+If this proves difficult, we can let implementers of the container choose
+whether to expose values or pointers to values. When electing to "pointers to
+values", an adapter can be provided to support transforming the range of
+pointers into a range a values. This choice allows supporting either iteration
+over container r-values (`ElementType = T`), or efficient iteration
+(`ElementType = T*`), but not both. While acceptable as an intermediate
+solution, a better alternative will be needed in the future.
+
+```carbon
+adapter IterateByValue(T:! Iterate where .CursorType impls Deref) for T {
+  impl as Iterate
+      where .CursorType = T.CursorType
+        and .ElementType = T.CursorType.(Deref.Result) {
+        ...
+  }
+}
+
+// Used like:
+var frames: NetworkFramesContainerByPtr = ...
+for (i: NetworkFrame in frames as IterateByValue(NetworkFramesContainerByPtr))
+{
+  ...
+}
+```
+
+Alternatively, we can allow the binding in the `for` loop to be an `addr`
+pattern, implemented by calling into a separate interface on the container. This
+puts the choice of using values or pointers to values in the hands of the `for`
+loop author. While this deviates from the ergonomics of `let` (where the
+compiler can choose to copy or alias the value depending on what is more
+efficient), this option would provide parity with C++ `for` loops (where the
+author chooses between value and reference).
+
+### C++ interoperability
+
+#### Iterating over C++ types in Carbon
+
+For reference, range-based `for` loops in C++ requires:
+
+-   `begin()` and `end()` methods or free functions, and
+-   the type returned supports pre-increment `++`, indirection `*`, and
+    inequality `!=` operations
+
+See section 8.6.4 of N4860 for more details.
```

Comment:
Can this be made into a link?

Reply:
Do we have any preferred source to link the standard in an easy to read way? This is what I have https://eel.is/c++draft/stmt.iter#stmt.ranged from the latest draft.

Reply:
I don't have a preferred source. The link you mention is fine with me. For non-spec C++ links, I tend to use cppreference.com

Reply:
Done